internal/xds/translator/ratelimit.go

```diff
@@ -80,50 +71,65 @@ func (t *Translator) isRateLimitPresent(irListener *ir.HTTPListener) bool {
 	}
 	// Return true if rate limit config exists.
 	for _, route := range irListener.Routes {
-		if routeContainsGlobalRateLimit(route) {
+		if isValidGlobalRateLimit(route) {
 			return true
 		}
 	}
 	return false
 }
 
 // buildRateLimitFilter constructs a list of HTTP filters for rate limiting based on the provided HTTP listener configuration.
-// It creates at most one filter per domain to avoid duplicates.
+// It creates separate filters for shared and non-shared rate limits.
 func (t *Translator) buildRateLimitFilter(irListener *ir.HTTPListener) []*hcmv3.HttpFilter {
 	if irListener == nil || irListener.Routes == nil {
 		return nil
 	}
 
-	var filters []*hcmv3.HttpFilter
-	// Map to track which domains we've already created filters for, prevents creating duplicate filters for the same domain
-	processedDomains := make(map[string]bool)
+	filters := []*hcmv3.HttpFilter{}
+	created := make(map[string]bool)
 
-	// Iterate over each route in the listener to create rate limit filters for shared rate limits.
 	for _, route := range irListener.Routes {
-		if !routeContainsGlobalRateLimit(route) {
+		if !isValidGlobalRateLimit(route) {
 			continue
 		}
 
-		var domain string
-		filterName := getRateLimitFilterName(route)
-		if isSharedRateLimit(route) {
-			// For shared rate limits, use the domain derived from the traffic policy
-			domain = getDomainName(route)
-		} else {
-			// For non-shared rate limits, use the listener domain
-			domain = irListener.Name
+		hasShared, hasNonShared := false, false
+		var sharedRuleName string
+
+		for _, rule := range route.Traffic.RateLimit.Global.Rules {
+			if isRuleShared(rule) {
+				hasShared = true
+				if sharedRuleName == "" {
```

Comment:
is `sharedRuleName == ""` required , or should we rely on `stripRuleIndexSuffix` to gracefully handle it even when the input is unlikely